### PR TITLE
ref(rules): Explicit Rule IDs

### PR DESCRIPTION
**Pull request by mgaeta**
We noticed a potentially catastrophic vulnerability in the way we define Rules. Rules are either "filters", "conditions", or "actions" and are implemented as classes in the Rules hierarchy. The are accessible through the `RuleRegistry`:
```py
class RuleRegistry:
    ...
    def add(self, rule):
        self._map[rule.id] = rule

    def get(self, rule_id, type=None):
        return self._map.get(rule_id)
```
The are accessed in a similarly obvious way:
```py
_SENTRY_RULES = (
    "sentry.mail.actions.NotifyEmailAction",
    ...
    "sentry.rules.filters.level.LevelFilter",
)

def init_registry() -> RuleRegistry:
    registry = RuleRegistry()
    for rule in _SENTRY_RULES:
        cls = import_string(rule)
        registry.add(cls)
    return registry
rules = init_registry()
```
The problem arises because of how the rule classes are assigned IDs:
```py
class RuleDescriptor(type):
    def __new__(cls, *args, **kwargs):
        new_cls = super().__new__(cls, *args, **kwargs)
        new_cls.id = f"{new_cls.__module__}.{new_cls.__name__}"
        return new_cls

class RuleBase(metaclass=RuleDescriptor):
    ...
```
This means that moving Rules between files or changing their names will change their IDs. 

Rules are are stored in the in the DB as a JSON blob:
```py
class Rule(Model):
    ...
    data = GzippedDictField()
```
The looks something like:
```json
{
  "actions": [...],
  "conditions": [{
    "id":  "sentry.rules.conditions.event_attribute.EventAttributeCondition",
    ...
  }],
  "filters": [],
}
```   
If the IDs of the rules ever change, then all the DB rules that reference them by ID will break.

For an example of how a name change could go undetected, see https://github.com/getsentry/sentry/pull/32512. We are somewhat confident that this hasn't already happened because we'd expect to see `warning` logs like:
```
Unregistered condition sentry.rules.conditions.event_attribute.EventAttributeCondition
```
This PR solves this problem by hardcoding the IDs as class variables on _every_ rule.

### CODE REVIEW

File: src/sentry/rules/base.py [hunk 1]
```diff
@@ -37,14 +37,7 @@
...
 
-class RuleDescriptor(type):
-    def __new__(cls, *args, **kwargs):
-        new_cls = super().__new__(cls, *args, **kwargs)
-        new_cls.id = f"{new_cls.__module__}.{new_cls.__name__}"
-        return new_cls
-
-
-class RuleBase(metaclass=RuleDescriptor):
+class RuleBase(abc.ABC):
```
Comment (wedamija) [inline src/sentry/rules/base.py:40]: Should `id` be an required property here?

Comment (mgaeta) [reply src/sentry/rules/base.py:40]: Python 3.9 has a nice pattern for this:
```py
@classmethod
@property
def id(cls) -> str:
    raise NotImplemented
```
I'm not sure how to do this in 3.8 without messing with metaclasses.

Comment (wedamija) [reply src/sentry/rules/base.py:40]: I think this works:
```
    @property
    @abstractmethod
    def my_abstract_property(self):
        ...
```

https://docs.python.org/3.8/library/abc.html#abc.abstractmethod

Comment (wedamija) [reply src/sentry/rules/base.py:40]: I guess it's not a class level property but I think that's still fine

File: src/sentry/rules/base.py [hunk 2]
```diff
@@ -56,6 +49,11 @@ def __init__(self, project, data=None, rule=None):
         self.had_data = data is not None
         self.rule = rule
 
+    @property
+    @abc.abstractmethod
+    def id(self) -> str:
+        raise NotImplementedError
```
Comment (wedamija) [inline src/sentry/rules/base.py:55]: You don't need to raise inside of an abstract method, `pass` is fine.

File: src/sentry/integrations/msteams/notify_action.py [hunk 3]
```diff
@@ -56,6 +56,7 @@ def clean(self) -> dict[str, Any] | None:
 
 
 class MsTeamsNotifyServiceAction(IntegrationEventAction):
+    id = "sentry.integrations.msteams.MsTeamsNotifyServiceAction"
```
Comment (wedamija) [inline src/sentry/integrations/msteams/notify_action.py:59]: This does not match the old ID

File: src/sentry/integrations/pagerduty/notify_action.py [hunk 4]
```diff
@@ -86,6 +86,7 @@ def clean(self) -> dict[str, Any] | None:
 
 
 class PagerDutyNotifyServiceAction(IntegrationEventAction):
+    id = "sentry.integrations.pagerduty.PagerDutyNotifyServiceAction"
```
Comment (wedamija) [inline src/sentry/integrations/pagerduty/notify_action.py:89]: Same as above

**Review by wedamija — CHANGES_REQUESTED** [review]
I added back the descriptor to test the IDs and some are wrong

```
class RuleDescriptor(type):
    def __new__(cls, *args, **kwargs):
        new_cls = super().__new__(cls, *args, **kwargs)
        gen_id = f"{new_cls.__module__}.{new_cls.__name__}"
        if new_cls.id != "":
            assert new_cls.id == gen_id, (new_cls.id, gen_id)
        print(f"set id {new_cls.id}, gen id {gen_id}")
        return new_cls


class RuleBase(metaclass=RuleDescriptor):
    label = None
    form_cls = None
    id = ""
```
Also had to remove the ABCs to get this to run, but this is a good way to audit your changes here.

**Review by wedamija — APPROVED** [review]
Please monitor this while it deploys. I think we've caught any potential problems, but want to be sure.

**Pull request merged**